core/src/components/modal/modal.tsx

```diff
@@ -650,6 +639,8 @@ export class Modal implements ComponentInterface, OverlayInterface {
       return false;
     }
 
+    const unlock = await this.lockController.lock();
```

Comment:
Can we add a comment on why this lock is before the `checkCanDismiss` call?

Reply:
We also need to release the lock inside of that `checkCanDismiss` if block

Reply:
I guess I'm not actually sure why it's before the `checkCanDismiss` call? It seems like we would need the lock to be claimed before any `await`s, for timing reasons. But beyond that, I'm not sure.

Reply:
Pretty much! The `canDismiss` check is async (since devs can pass a promise), so the dismiss transition can _potentially_ run but we won't know until that canDismiss callback resolves, so we claim the lock just in case.

If we claimed the lock _after_ that `checkCanDismiss` call and then opened, closed, and then re-opened the modal then the modal would not re-open.

Reply:
Here's an example you can try:

```html
<script>
  const modal = document.querySelector('ion-modal');
  modal.canDismiss = () => {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve(true);
      }, 2000);
    })
  }

  setTimeout(() => {
    modal.isOpen = true;
    setTimeout(() => {
      modal.isOpen = false;
      setTimeout(() => {
        modal.isOpen = true;
      }, 50);
    }, 50);
  }, 1000);
</script>
```

| lock is claimed before `checkCanDismiss` | lock is claimed after `checkCanDismiss` | 
| - | - |
| <video src="https://github.com/ionic-team/ionic-framework/assets/2721089/eb623850-b6a4-4ffe-8f78-5fb95a0c1f0c"></video> | <video src="https://github.com/ionic-team/ionic-framework/assets/2721089/0fa2a8dd-d982-4334-ad5b-20a3771c02fa"></video> |